modules/codegen/src/test/scala/tests/core/TypesTest.scala

```diff
@@ -320,4 +321,60 @@ class TypesTest extends AnyFunSuite with Matchers with SwaggerSpecRunner {
     cls.structure shouldEqual definition.structure
     cmp.structure shouldEqual companion.structure
   }
+
+  test("Optional fields works") {
+    val swagger =
+      s"""
+         |swagger: "2.0"
+         |info:
+         |  title: Whatever
+         |  version: 1.0.0
+         |host: localhost:1234
+         |definitions:
+         |  TestObject:
+         |    type: object
+         |    required:
+         |      - required
+         |      - required-nullable
+         |    properties:
+         |      required:
+         |        type: string
+         |      required-nullable:
+         |        type: string
+         |        x-nullable: true
+         |      optional:
+         |        type: string
+         |        x-nullable: false
+         |      optional-nullable:
+         |        type: string
+         |        x-nullable: true
+         |      legacy:
+         |        type: string
+         |""".stripMargin
+    val (
+      ProtocolDefinitions(ClassDefinition(_, _, _, cls, staticDefns, _) :: Nil, _, _, _),
+      _,
+      _
+    ) = runSwaggerSpec(swagger)(Context.empty, Http4s)
+
+    val cmp = companionForStaticDefns(staticDefns)
+
+    val companion =
+      q"""
+        object TestObject {
+          implicit val encodeTestObject: Encoder.AsObject[TestObject] = {
+            val readOnlyKeys = Set[String]()
+            Encoder.AsObject.instance[TestObject](a => JsonObject.fromIterable(Vector(("required", a.required.asJson), ("required-nullable", a.requiredNullable.asJson), ("legacy", a.legacy.asJson)) ++ a.optional.map(value => ("optional", value.asJson)) ++ a.optionalNullable.fold(ifAbsent = None, ifPresent = value => Some("optional-nullable" -> value.asJson)))).mapJsonObject(_.filterKeys(key => !(readOnlyKeys contains key)))
+          }
+          implicit val decodeTestObject: Decoder[TestObject] = new Decoder[TestObject] { final def apply(c: HCursor): Decoder.Result[TestObject] = for (v0 <- c.downField("required").as[String]; v1 <- c.downField("required-nullable").as[Json].flatMap(_.as[Option[String]]); v2 <- c.downField("optional").as[Json].map(_.as[String].map(Some(_))).getOrElse(Right(None)); v3 <- c.downField("optional-nullable").as[Json].map(_.as[Option[String]].map(Property.Present(_))).getOrElse(Right(Property.Absent)); v4 <- c.downField("legacy").as[Option[String]]) yield TestObject(v0, v1, v2, v3, v4) }
```

Comment:
Question, should `required-nullable` be `Property[String]` instead of `Option[String]`? This seems asymmetric, compared to `optionalNullable: Property[Option[String]]`

Reply:
Not sure if I understood correctly, but...are you suggesting to add a third case to `Property`? So we have cases such as `Absent`, `Present` and `Value`?

This was actually my very initial implementation, but I ended up with this instead. Current implementaiton seemed to be easier to use, but now I'm having some second thoughts. Especially when I consider that `T` in `Property is actually always expected to be `Option`. So I would actually prefer to change this.

Let me know what you think.

Reply:
Well not even a third case, just `Property[String]` with the possible values of `Absent` and `Present(value)`. This is consistent with how `optional-nullable` represents `Present(Some(value))`/`Present(None)`/`Absent`.

If we think in terms of an imaginary world where `sealed trait Property[F[_], +A]` would be fun to work with,

| optional-nullable   | required-nullable |
| ------------------- | ----------------- |
| Property[Option, A] | Property[Id, A]   |

we get a consistent model.

Reply:
Ok, I think I understand.  You essentially propose to change `Present` case into `final case class Present[+T](value: Option[T]) extends Property[T]`. Is that correct?

Reply:
Sorry, must have miscommunicated -- I'm thinking along the lines of:
```diff
 implicit val decodeTestObject: Decoder[TestObject] = new Decoder[TestObject] {
   final def apply(c: HCursor): Decoder.Result[TestObject] = for {
     v0 <- c.downField("required").as[String]
-    v1 <- c.downField("required-nullable").as[Json].flatMap(_.as[Option[String]])
+    v1 <- c.downField("required-nullable").as[Json].flatMap(_.as[Option[String]].map(_.map(Property.Present(_)).getOrElse(Property.Absent)))
     v2 <- c.downField("optional").as[Json].map(_.as[String].map(Some(_))).getOrElse(Right(None))
     v3 <- c.downField("optional-nullable").as[Json].map(_.as[Option[String]].map(Property.Present(_))).getOrElse(Right(Property.Absent))
     v4 <- c.downField("legacy").as[Option[String]]
   } yield TestObject(v0, v1, v2, v3, v4)
 }
```

We're already using `Property[Option[T]]` for `optional-nullable`, so it's not too much to generate another case with this type.

Reply:
I'm realizing that I may be confusing the intent of `Property` to describe the `Nullability` of a value, which may not have been how you intended it to be used when writing this

Reply:
However it makes sense to use `Property` for `optional` case. It would make it somehow consistent. `Property` is used to describe presence/absent of the field in the json, while `Option` is used to `Nullability`.

Reply:
Changed!

Reply:
i think name `Property` is a bit confusing and too generic ... could we name it something like `Nullable`? i think that is much more descriptive

Reply:
Nullability is handled by `Option` data type. `Property` refers to a presence of a key in a json.

Reply:
ah, i think i understand now ... i still think that `Property` name is too generic - for trait `Property` I would expect that the inheriting case classes would be things like `StringProperty, ObjectProperty, IntProperty etc...`

i would propose to rename it to something like `PropertyPresence`

Reply:
Could even go for `Presence.Present` vs `Presence.Absent`, in your other comment you indicated that this will end up leaking into user code, a good type name is important here.

Additionally, I think we establish this pattern in a few other places, but we will want to partially or fully quality the places we use this type. Additions like this into the common namespace of what can be generated in a specification reduce the set of specifications that don't need to be modified to work out of the box with guardrail.

Reply:
Renamed to `Presence` and it is always referenced using FQN from DTO classes (even though I belieive there might still be issues with name clashes)